### ISSUE TITLE
[BUG]: Extracted Embeddings

### ISSUE THREAD

**Issue by Suhail-SSRC** [serengil/retinaface#114]
### Before You Report a Bug, Please Confirm You Have Done The Following...

- [X] I have updated to the latest version of the packages.
- [X] I have searched for both [existing issues](https://github.com/serengil/retinaface/issues) and [closed issues](https://github.com/serengil/retinaface/issues?q=is%3Aissue+is%3Aclosed) and found none that matched my issue.

### RetinaFace's Version

v0.0.17

### Python Version

3.9.20

### Operating System

Ubuntu 22.04

### Dependencies

absl-py==2.1.0
anyio==4.6.2.post1
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
array_record==0.5.1
arrow==1.3.0
asttokens==2.4.1
astunparse==1.6.3
async-lru==2.0.4
attrs==24.2.0
babel==2.16.0
beautifulsoup4==4.12.3
bleach==6.2.0
cachetools==5.5.0
certifi==2024.8.30
cffi==1.17.1
charset-normalizer==3.4.0
click==8.1.7
comm==0.2.2
contourpy==1.3.0
cycler==0.12.1
debugpy==1.8.9
decorator==5.1.1
defusedxml==0.7.1
dm-tree==0.1.8
etils==1.5.2
exceptiongroup==1.2.2
executing==2.1.0
fastjsonschema==2.20.0
filelock==3.16.1
flatbuffers==24.3.25
fonttools==4.55.0
fqdn==1.5.1
fsspec==2024.10.0
ftfy==6.3.1
gast==0.4.0
gdown==5.2.0
glob2==0.7
google-auth==2.36.0
google-auth-oauthlib==1.0.0
google-pasta==0.2.0
graphviz==0.8.4
grpcio==1.68.0
h11==0.14.0
h5py==3.12.1
httpcore==1.0.7
httpx==0.27.2
idna==3.10
imageio==2.36.0
importlib_metadata==8.5.0
importlib_resources==6.4.5
ipykernel==6.29.5
ipython==8.18.1
ipywidgets==8.1.5
isoduration==20.11.0
jax==0.4.30
jaxlib==0.4.30
jedi==0.19.2
Jinja2==3.1.4
joblib==1.4.2
json5==0.9.28
jsonpointer==3.0.0
jsonschema==4.23.0
jsonschema-specifications==2024.10.1
jupyter==1.1.1
jupyter-console==6.6.3
jupyter-events==0.10.0
jupyter-lsp==2.2.5
jupyter_client==8.6.3
jupyter_core==5.7.2
jupyter_server==2.14.2
jupyter_server_terminals==0.5.3
jupyterlab==4.2.6
jupyterlab_pygments==0.3.0
jupyterlab_server==2.27.3
jupyterlab_widgets==3.0.13
keras==2.12.0
keras-cv-attention-models==1.4.1
kiwisolver==1.4.7
lazy_loader==0.4
libclang==18.1.1
Markdown==3.7
markdown-it-py==3.0.0
MarkupSafe==3.0.2
matplotlib==3.9.2
matplotlib-inline==0.1.7
mdurl==0.1.2
mistune==3.0.2
ml-dtypes==0.4.1
mxnet==1.9.1
namex==0.0.8
nbclient==0.10.0
nbconvert==7.16.4
nbformat==5.10.4
nest-asyncio==1.6.0
networkx==3.2.1
notebook==7.2.2
notebook_shim==0.2.4
numpy==1.23.5
nvidia-cublas-cu11==11.11.3.6
nvidia-cudnn-cu11==8.6.0.163
oauthlib==3.2.2
opencv-python==4.10.0.84
opt_einsum==3.4.0
optree==0.13.1
overrides==7.7.0
packaging==24.2
pandas==2.2.3
pandocfilters==1.5.1
parso==0.8.4
pexpect==4.9.0
pillow==11.0.0
platformdirs==4.3.6
prometheus_client==0.21.0
promise==2.3
prompt_toolkit==3.0.48
protobuf==3.20.3
psutil==6.1.0
ptyprocess==0.7.0
pure_eval==0.2.3
pyasn1==0.6.1
pyasn1_modules==0.4.1
pycparser==2.22
Pygments==2.18.0
pyparsing==3.2.0
PySocks==1.7.1
python-dateutil==2.9.0.post0
python-json-logger==2.0.7
pytz==2024.2
PyYAML==6.0.2
pyzmq==26.2.0
referencing==0.35.1
regex==2024.11.6
requests==2.32.3
requests-oauthlib==2.0.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rich==13.9.4
rpds-py==0.21.0
rsa==4.9
scikit-image==0.24.0
scikit-learn==1.5.2
scipy==1.13.1
Send2Trash==1.8.3
six==1.16.0
sniffio==1.3.1
soupsieve==2.6
stack-data==0.6.3
tensorboard==2.12.3
tensorboard-data-server==0.7.2
tensorflow==2.12.0
tensorflow-datasets==4.9.3
tensorflow-estimator==2.12.0
tensorflow-io-gcs-filesystem==0.37.1
tensorflow-metadata==1.16.1
termcolor==2.5.0
terminado==0.18.1
threadpoolctl==3.5.0
tifffile==2024.8.30
tinycss2==1.4.0
toml==0.10.2
tomli==2.1.0
tornado==6.4.2
tqdm==4.67.1
traitlets==5.14.3
types-python-dateutil==2.9.0.20241003
typing_extensions==4.12.2
tzdata==2024.2
uri-template==1.3.0
urllib3==2.2.3
wcwidth==0.2.13
webcolors==24.11.1
webencodings==0.5.1
websocket-client==1.8.0
Werkzeug==3.1.3
widgetsnbextension==4.0.13
wrapt==1.14.1
zipp==3.21.0

### Reproducible example

```Python
import os
import cv2
import numpy as np
import matplotlib.pyplot as plt
from sklearn.manifold import TSNE
from tqdm import tqdm
from face_detector import RetinaFaceV1
import tensorflow as tf

def process_dataset(dataset_path, recognizer_model_path):
    """
    Process all images in a dataset to extract embeddings and visualize them using t-SNE.

    Args:
        dataset_path (str): Path to the dataset folder containing subfolders of class images.
        recognizer_model_path (str): Path to the recognizer model for embedding extraction.

    Returns:
        None
    """
    # Initialize the RetinaFace detector and the recognizer model
    face_detector = RetinaFaceV1()
    recognizer_model = tf.keras.models.load_model(recognizer_model_path, compile=False)

    # Prepare arrays to store embeddings and labels
    all_embeddings = []
    all_labels = []

    print(f"Processing dataset at: {dataset_path}")
    for class_folder in tqdm(os.listdir(dataset_path), desc="Classes"):
        class_path = os.path.join(dataset_path, class_folder)
        if not os.path.isdir(class_path):
            continue  # Skip files, process only folders (classes)

        for image_name in os.listdir(class_path):
            image_path = os.path.join(class_path, image_name)
            frame = cv2.imread(image_path)
            if frame is None:
                print(f"Warning: Could not read {image_path}. Skipping.")
                continue

            # Detect faces
            bbs, _, _ = face_detector.detect_in_image(frame)
            aligned_faces = face_detector.extract_aligned_faces(frame)

            if len(aligned_faces) > 0:
                # Extract embeddings for all detected faces
                embeddings = recognizer_model((aligned_faces - 127.5) * 0.0078125).numpy()
                all_embeddings.extend(embeddings)
                all_labels.extend([class_folder] * len(embeddings))  # Label each embedding with the class name

    # Convert to numpy arrays
    all_embeddings = np.array(all_embeddings)
    all_labels = np.array(all_labels)

    # Perform t-SNE for visualization
    print("Performing t-SNE visualization...")
    tsne = TSNE(n_components=2, random_state=42)
    reduced_embeddings = tsne.fit_transform(all_embeddings)

    # Plot t-SNE results
    plt.figure(figsize=(12, 8))
    unique_labels = np.unique(all_labels)
    for label in unique_labels:
        indices = all_labels == label
        plt.scatter(reduced_embeddings[indices, 0], reduced_embeddings[indices, 1], label=label, alpha=0.7)
    plt.legend(loc="upper left", bbox_to_anchor=(1, 1), title="Classes")
    plt.title("t-SNE Visualization of Embeddings")
    plt.xlabel("t-SNE Dimension 1")
    plt.ylabel("t-SNE Dimension 2")
    plt.tight_layout()
    plt.savefig("tsne_visualization_retina_embd.png", dpi=300)
    plt.show()

    print("t-SNE visualization saved as 'tsne_visualization.png'.")


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Process a dataset with RetinaFace and visualize embeddings using t-SNE.")
    parser.add_argument("-d", "--dataset_path", type=str, required=True, help="Path to the dataset folder.")
    parser.add_argument("-r", "--recognizer_model", type=str, required=True, help="Path to the recognizer model file.")
    args = parser.parse_args()

    process_dataset(args.dataset_path, args.recognizer_model)
```

### Expected Result

Wehn I am using YOLO, and GhostFaceNet. The seperation between classes is clear. Retina has a high detection rate but it might be the way the faces are extracted or alligned
![tsne_visualization_retina_embd](https://github.com/user-attachments/assets/56ec5213-2fe2-4bbd-8238-c813c6d01e23)
 
![svm_embeddings_visualization](https://github.com/user-attachments/assets/c705d6c0-ba96-4496-b7af-7f8256084899)

### Additional Info

Download GhostFaceNet pretrained model from here: https://github.com/HamadYA/GhostFaceNets/releases/download/v1.2/GhostFaceNet_W1.3_S1_ArcFace.h5

**Comment by serengil** [comment]
What is the bug here? What do you expect?

**Comment by Suhail-SSRC** [comment]
I expect when passing it to the recognizer, the separation between classes should be more unique, just like the YOLO (The separated embeddings in the second picture)

**Comment by serengil** [comment]
That is not responsibility of retinaface. It is just responsible for extracting faces.
